eos-paygd/main.c

```diff
@@ -194,24 +240,52 @@ main (int   argc,
       watchdog_id = g_timeout_add_seconds_full (G_PRIORITY_HIGH, 10, ping_watchdog, NULL, NULL);
       g_assert (watchdog_id > 0);
 
-      /* Let systemd know it's okay to proceed to pivot to the final root */
-      sd_notify_ret = sd_notify (0, "READY=1");
+      /* Here be dragons:
+       * We're currently in the initramfs root directory, and systemd is putting
+       * all the useful bits of the system into /sysroot in preparation for the
+       * root pivot. After the root pivot, it will delete everything in the
+       * initramfs root, which will leave us in limbo forever.  We need to
+       * chroot() out of this into sysroot.
+       *
+       * However, once we do that, we no longer have access to the socket we
+       * need to write to to inform systemd we're ready for it to do the root
+       * pivot.
+       *
+       * What we do is chdir() into the directory $NOTIFY_SOCKET resides in,
+       * chroot() into /sysroot, then use our own variant of sd_notify() that
+       * doesn't refuse to use relative paths to notify systemd we're ready.
+       *
+       * After that, we can chdir() into our new / and begin trying to connect
+       * to the dbus socket. */
+
+      sd_socket_env = g_getenv("NOTIFY_SOCKET");
+      if (!sd_socket_env)
+        {
+          /* If we can't notify systemd that we're ready to move on, it will
+           * timeout in 90 seconds and shutdown, might as well just exit. */
+          g_printerr ("%s: NOTIFY_SOCKET not set.\n", argv[0]);
+          return SD_NOTIFY_FAILURE_EXIT_CODE;
+        }
+      sd_socket_dir = g_path_get_dirname (sd_socket_env);
+      sd_socket_name = g_path_get_basename (sd_socket_env);
+      if (!sd_socket_dir || !sd_socket_name)
+        {
+          g_printerr ("%s: NOTIFY_SOCKET not in valid format.\n", argv[0]);
+          return SD_NOTIFY_FAILURE_EXIT_CODE;
+        }
+      if (chdir (sd_socket_dir))
+        g_warning ("Unable to change working dir to systemd socket dir (%s): %m", sd_socket_dir);
+      if (chroot ("/sysroot"))
+        g_warning ("Unable to switch root to run-time root directory (/sysroot): %m");
+
+      sd_notify_ret = payg_relative_sd_notify (sd_socket_name, "READY=1");
```

Comment:
Is the socket we're using here the same one as in the initrd filesystem? And if not, how do we know systemd will have already created it in the new sysroot?

It's possible that's a stupid question because I don't have a good conceptual grasp on how sockets work.

Reply:
yes, the socket we're using is the one in initramfs (exactly what systemd told us to use)

essentially something like this:
NOTIFY_SOCKET=/var/run/systemd/notify
we cd /var/run/systemd
we use "notify" as the socket name.

Reply:
Okay. It feels a bit fragile since maybe a process having a cwd outside its root directory is a rare situation that may be broken by a kernel update. But I don't have a better idea.